lib/UserCohort.js

```diff
@@ -0,0 +1,67 @@
+const mongoose = require("./utils/mongoose");
+const Schema = mongoose.Schema;
+const User = require("./User");
+
+const UserCohort = new Schema(
+  {
+    user: {
+      type: Schema.Types.ObjectId,
+      ref: "User",
+      required: "{PATH} is required!",
+      unique: true
+    },
+    cohort: { type: String }
+  },
+  { timestamps: true }
+);
+UserCohort.plugin(require("./plugins/mongoose-to-id"));
+UserCohort.plugin(require("mongoose-findorcreate"));
+UserCohort.plugin(require("./plugins/mongoose-no-underscore-id"));
+
+UserCohort.statics.findForUser = async function(user) {
+  if (!(user instanceof User)) {
+    throw new Error("user param must be a User instance");
+  }
+  return await this.findOne({ user: user.id }).exec();
+};
+
+UserCohort.statics.setUserCohort = async function(user, cohortName) {
+  if (!(user instanceof User)) {
+    throw new Error("user param must be a User instance");
+  }
+
+  const UserCohortModel = this;
+  let UserCohort = await UserCohortModel.findForUser(user);
+  if (UserCohort && UserCohort.cohort === cohortName) {
+    return UserCohort;
+  }
+
+  UserCohort = await UserCohortModel.findOneAndUpdate(
+    { user: user.id },
+    {
+      user: user.id,
+      cohort: cohortName
+    },
+    { upsert: true, new: true }
+  ).exec();
+
+  // Check if user is in any cohorts
+  const goalCohortModel = mongoose.model("GoalCohort");
+  const cohorts = await goalCohortModel.find({
+    members: {
+      $elemMatch: {
+        user: user.id
+      }
+    }
+  });
+
+  // Leave old cohorts
+  const goalModel = mongoose.model("Goal");
+  for (const cohort of cohorts) {
+    const goal = await goalModel.findById(cohort.goal);
+    await goalCohortModel.leaveCohort(user, goal);
```

Comment:
I think I requested that this be compressed into a single `delete all` type query earlier, but on thinking about it more, I think you have it right. The business logic for removing a user from a goal cohort is too complicated to do it globally